### PR TITLE
[PF-2791] Change serdes for state exceptions; remove exception from DeleteFailure

### PR DESCRIPTION
**Pull request by ddietterich**
My code to serialize and deserialize exceptions on state changes didn't work. This change adds a method of normalizing the exceptions stored. Added another method that tolerate column contents that cannot be deserialized, rather than breaking workspace retrieval altogether.

Also, this removes the ability to set any exception on a workspace delete failure. That shouldn't have been there in the first place.

### CODE REVIEW

File: service/src/main/java/bio/terra/workspace/db/StateDao.java [hunk 1]
```diff
@@ -121,6 +113,51 @@ public void updateState(
         errorReport);
   }
 
+  @VisibleForTesting
+  public String normalizeException(Exception exception) {
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/db/StateDao.java:117]: @nullable

File: service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceUnitTest.java [hunk 2]
```diff
@@ -31,17 +35,39 @@
...
+        new ForbiddenException(
+            "User mnemosyne.ninetynine@gmail.com is not authorized to perform action delete on controlled-application-private-workspace-resource ef5ee667-48f1-4407-8ca1-0efe109591c7");
+    String errorJson = stateDao.normalizeException(fex);
+
+    var ex = StateDao.deserializeException(errorJson);
+    assertEquals(fex.getStatusCode(), ex.getStatusCode());
+    assertTrue(StringUtils.contains(ex.getMessage(), fex.getMessage()));
+
+    // Case where junk is in the exception
+    var exrex = StateDao.deserializeException("bad bad bad");
```
Comment (yuhuyoyo) [inline service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceUnitTest.java:67]: split this into separate test please

File: service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/workspace/RunDeleteCloudContextFlightStep.java [hunk 3]
```diff
@@ -98,12 +99,26 @@ public StepResult undoStep(FlightContext context) throws InterruptedException {
...
+        "Delete {} cloud context state is {}", cloudPlatform, flightState.getFlightStatus());
     return (flightState.getFlightStatus() == FlightStatus.ERROR
         || flightState.getFlightStatus() == FlightStatus.FATAL
         || flightState.getFlightStatus() == FlightStatus.SUCCESS);
   }
+
+  private String getFlightId(FlightContext context, CloudPlatform cloudPlatform) {
+    // Retrieve our flight id
+    Map<CloudPlatform, String> flightIds =
+        context.getWorkingMap().get(WorkspaceFlightMapKeys.FLIGHT_IDS, new TypeReference<>() {});
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/workspace/RunDeleteCloudContextFlightStep.java:111]: getrequred

**Review by yuhuyoyo — APPROVED** [review]

File: service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/workspace/MakeFlightIdsStep.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  public MakeFlightIdsStep(UUID workspaceUuid, WorkspaceDao workspaceDao) {
+    this.workspaceUuid = workspaceUuid;
+    this.workspaceDao = workspaceDao;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    Map<CloudPlatform, String> flightIds = new HashMap<>();
+
+    // For each cloud context in the workspace, run the cloud context delete flight
```
Comment (zloery) [inline service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/workspace/MakeFlightIdsStep.java:35]: update/delete comment, this is just generating the IDs, not running the flight

**Review by zloery — APPROVED** [review]

File: service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/workspace/MakeFlightIdsStep.java (hunk 4 above)
Comment (ddietterich) [reply service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/workspace/MakeFlightIdsStep.java:35]: Cut'n'paste always gets me in trouble. I removed the comment

File: service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/workspace/RunDeleteCloudContextFlightStep.java (hunk 3 above)
Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/workspace/RunDeleteCloudContextFlightStep.java:111]: oh sorry didn't know that there isn't one.

Comment (ddietterich) [reply service/src/main/java/bio/terra/workspace/service/workspace/flight/delete/workspace/RunDeleteCloudContextFlightStep.java]: NP 
It is trivial.

**Pull request merged**